micropip/package_index.py

```diff
@@ -0,0 +1,188 @@
+import sys
+from collections import defaultdict
+from collections.abc import Generator
+from dataclasses import dataclass
+from typing import Any
+
+from packaging.utils import InvalidWheelFilename
+from packaging.version import InvalidVersion, Version
+
+from ._utils import is_package_compatible, parse_version
+
+
+# TODO: Merge this class with WheelInfo
+@dataclass
+class ProjectInfoFile:
+    filename: str  # Name of the file
+    url: str  # URL to download the file
+    version: Version  # Version of the package
+    sha256: str  # SHA256 hash of the file
+
+    # Size of the file in bytes, if available (PEP 700)
+    # This key is not available in the Simple API HTML response, so this field may be None
+    size: int | None = None
+
+
+@dataclass
+class ProjectInfo:
+    """
+    This class stores common metadata that can be obtained from different APIs (JSON, Simple)
+    provided by PyPI. Responses received from PyPI or other package indexes that support the
+    same APIs must be converted to this class before being processed by micropip.
+    """
+
+    name: str  # Name of the package
+
+    # List of releases available for the package, sorted in ascending order by version.
+    # For each version, list of wheels compatible with the current platform are stored.
+    # If no such wheel is available, the list is empty.
+    releases: dict[Version, Generator[ProjectInfoFile, None, None]]
+
+    @staticmethod
+    def from_json_api(data: dict[str, Any]) -> "ProjectInfo":
+        """
+        Parse JSON API response
+
+        https://warehouse.pypa.io/api-reference/json.html
+        """
+
+        name: str = data.get("info", {}).get("name", "UNKNOWN")
+        releases_raw: dict[str, list[Any]] = data["releases"]
+
+        # Filter out non PEP 440 compliant versions
+        releases: dict[Version, list[Any]] = {}
+        for version_str, fileinfo in releases_raw.items():
+            try:
+                version = Version(version_str)
+                if str(version) != version_str:
+                    continue
```

Comment:
Do we need to keep this? As long as there is a version it's fine no? And str(version) has a non negligible runtime.

Reply:
Good point. I think it is okay to remove it now.

That check was introduced at https://github.com/pyodide/pyodide/issues/2752, but I think it cannot happen on wheel files (unless someone modifies the version manually).